File: frontend/src/api/studio.ts
/**
 * OmicStudio AI 分析工作台 API
 *
 * 对应后端 /api/v1/studio 路由（P0-2）：
 * 会话创建/列表/详情、工作区文件、产物清单与下载。
 * 聊天流复用 POST /chat/stream（见 useAgentChatStream），
 * 代码重跑 SSE 见 @/composables/useStudioRunStream。
 */
import apiClient from './client'

function triggerBlobDownload(url: string, filename: string): void {
  const anchor = document.createElement('a')
  anchor.href = url
  anchor.download = filename
  anchor.style.display = 'none'
  document.body.appendChild(anchor)
  anchor.click()
  anchor.remove()

  window.setTimeout(() => URL.revokeObjectURL(url), 1_000)
}

/** Studio 会话（后端 ChatSessionDTO，mode="studio"） */
export interface StudioSessionDTO {
  session_id: string
  title: string
  agent_id: string | null
  model_id: string
  message_count: number
  status: string
  mode: 'studio'
  created_at: string
  updated_at: string
  last_message_at: string | null
  multi_agent?: boolean
  overdrive?: boolean
}

export type SandboxStatus = 'running' | 'stopped' | 'absent' | 'unavailable'

export interface WorkspaceEntry {
  name: string
  type: 'file' | 'dir'
  size: number
  mtime: number
}

export interface StudioCapabilityAuditEvent {
  action: 'list' | 'load' | string
  kind: 'skill' | 'mcp' | null
  capability_id: string | null
  success: boolean
  at: string
}

export interface StudioCapabilitiesState {
  loaded_skill_ids: string[]
  loaded_mcp_ids: string[]
  audit: StudioCapabilityAuditEvent[]
}


export interface ExtractStudioSkillPayload {
  path: string
  skill_id: string
  name: string
  description: string
  usage_instructions: string
  category: string
  icon: string
  confirm_no_secrets: boolean
}

export interface ExtractedSkill {
  id: string
  skill_id: string
  name: string
  description: string
  category: string
  is_active: boolean
}

export interface StudioShareResponse {
  token: string
  share_path: string
  expires_at: string
}

export interface StudioShareStatus {
  active: boolean
  expires_at: string | null
  shared_at: string | null
}

export interface SharedStudioMessage {
  role: 'user' | 'assistant'
  content: string
  created_at: string
}

export interface SharedStudioSession {
  title: string
  agent_id: string | null
  created_at: string
  updated_at: string
  expires_at: string
  messages: SharedStudioMessage[]
  artifacts: StudioArtifact[]
}

/** Studio 会话级权限模式（supervised=关键操作需用户批准；auto=放权自动执行） */
export type StudioPermissionMode = 'supervised' | 'auto'

export interface StudioPermissions {
  mode: StudioPermissionMode
}

/** 待审批的工具审批记录（GET /studio/approvals，对应 Redis 中的 pending 记录） */
export interface StudioPendingApproval {
  approval_id: string
  session_id: string
  tool_call_id: string
  tool_name: string
  arguments: Record<string, unknown>
  risk_hint: string
  status: 'pending'
  /** 剩余可审批秒数（Redis TTL） */
  expires_in: number
}

export interface StudioSessionDetail {
  session: StudioSessionDTO
  sandbox_status: SandboxStatus
  permissions?: StudioPermissions
  workspace_id: string | null
  files: WorkspaceEntry[]
  plan: StudioPlan | null
  capabilities: StudioCapabilitiesState | null
  sandbox_metrics: {
    cpu_percent: number
    memory_percent: number
    memory_used: number
    memory_limit: number
  }
  ui: {
    view_mode: 'chat' | 'split' | 'code'
    split_ratio: number
    follow_ai: boolean
    terminal_collapsed: boolean
    hibernate_on_leave: boolean
  }
  share: StudioShareStatus | null
}

export type StudioPlanStatus = 'pending' | 'in_progress' | 'done'

export interface StudioPlanStep {
  title: string
  status: StudioPlanStatus
}

export interface StudioPlan {
  steps: StudioPlanStep[]
}

export interface StudioArtifact {
  path: string
  size: number
  mtime: number
  download_url?: string
}

export interface StudioRunArtifact {
  path: string
  size: number
  mtime: number
}

export interface StudioImportResult {
  sandbox_path: string
  name: string
  size: number
  file_type: string
  input_files: string[]
}

export interface StudioFileEditResult {
  path: string
  diff: string
  size: number
}

export interface StudioArtifactRegistration {
  report_id: string
  title: string
  version: number
  parent_id: string | null
  file: Record<string, unknown>
}

/** 产物下载地址（需携带 JWT，前端统一走 blob 下载，见 downloadArtifact） */
export function artifactDownloadUrl(sessionId: string, path: string): string {
  return `/api/v1/studio/sessions/${sessionId}/artifacts/download?path=${encodeURIComponent(path)}`
}

export const studioApi = {
  /** 创建 Studio 会话（mode="studio"，workspace_id = session id） */
  async createSession(payload: {
    agent_id: string
    title?: string
    model_id?: string
  }): Promise<StudioSessionDTO> {
    const res = await apiClient.post<StudioSessionDTO>('/studio/sessions', payload)
    return res.data
  },

  /** 将普通聊天会话升级为 Studio 工作台（保留历史消息），可选重绑定 Agent */
  async promoteSession(sessionId: string, agentId?: string): Promise<StudioSessionDTO> {
    const res = await apiClient.post<StudioSessionDTO>(
      `/studio/sessions/${sessionId}/promote`,
      agentId ? { agent_id: agentId } : {},
    )
    return res.data
  },

  /** 从结果报告创建 Studio 会话并注入来源上下文 */
  async createSessionFromReport(payload: {
    report_id: string
    agent_id: string
  }): Promise<StudioSessionDTO> {
    const res = await apiClient.post<StudioSessionDTO>('/studio/sessions/from-report', payload)
    return res.data
  },

  /** 当前用户的 Studio 会话列表（按更新时间倒序） */
  async listSessions(): Promise<StudioSessionDTO[]> {
    const res = await apiClient.get<StudioSessionDTO[]>('/studio/sessions')
    return res.data || []
  },

  /** 会话详情 + 沙盒状态 + 工作区根目录条目 */
  async getSession(sessionId: string): Promise<StudioSessionDetail> {
    const res = await apiClient.get<StudioSessionDetail>(`/studio/sessions/${sessionId}`)
    return res.data
  },

  /** 更新会话级权限模式（supervised/auto，P1 HITL） */
  async updatePermissions(
    sessionId: string,
    mode: StudioPermissionMode,
  ): Promise<StudioPermissions> {
    const res = await apiClient.post<StudioPermissions>(
      `/studio/sessions/${sessionId}/permissions`,
      { mode },
    )
    // 后端返回 {permissions: {mode}} 或直接 {mode}，两种封装都兼容
    const data = res.data as StudioPermissions & { permissions?: StudioPermissions }
    return data.permissions || data
  },

  /** 列出指定会话仍处于 pending 的工具审批（刷新/断流重进会话时重建审批卡） */
  async listPendingApprovals(sessionId: string): Promise<StudioPendingApproval[]> {
    const res = await apiClient.get<{ approvals: StudioPendingApproval[] }>(
      '/studio/approvals',
      { params: { session_id: sessionId } },
    )
    return res.data?.approvals || []
  },

  /** 批准待审批的工具调用；带 modifiedArgs 即"编辑后批准" */
  async approveApproval(
    approvalId: string,
    modifiedArgs?: Record<string, unknown>,
  ): Promise<void> {
    await apiClient.post(`/studio/approvals/${approvalId}/approve`, {
      ...(modifiedArgs ? { modified_args: modifiedArgs } : {}),
    })
  },

  /** 退回待审批的工具调用（理由回灌 LLM 让其修改后重试） */
  async rejectApproval(approvalId: string, reason?: string): Promise<void> {
    await apiClient.post(`/studio/approvals/${approvalId}/reject`, {
      ...(reason ? { reason } : {}),
    })
  },

  async updateUi(
    sessionId: string,
    payload: { view_mode?: 'chat' | 'split' | 'code'; split_ratio?: number; follow_ai?: boolean; terminal_collapsed?: boolean },
  ): Promise<void> {
    await apiClient.patch(`/studio/sessions/${sessionId}/ui`, payload)
  },

  /** 释放沙盒容器但保留工作区；下一次执行会自动重建并挂回原目录。 */
  async hibernateSession(
    sessionId: string,
  ): Promise<{ status: 'hibernated' | 'busy' | 'absent' | 'unavailable'; workspace_preserved: boolean }> {
    const token = localStorage.getItem('access_token')
    const response = await fetch(`/api/v1/studio/sessions/${sessionId}/sandbox/hibernate`, {
      method: 'POST',
      headers: token ? { Authorization: `Bearer ${token}` } : {},
      keepalive: true,
    })
    if (!response.ok) throw new Error(`沙盒休眠失败 (${response.status})`)
    return response.json()
  },

  /** 创建或轮换只读分享令牌 */
  async createShare(sessionId: string, expiresHours = 168): Promise<StudioShareResponse> {
    const res = await apiClient.post<StudioShareResponse>(`/studio/sessions/${sessionId}/share`, {
      expires_hours: expiresHours,
    })
    return res.data
  },

  async getShareStatus(sessionId: string): Promise<StudioShareStatus> {
    const res = await apiClient.get<StudioShareStatus>(`/studio/sessions/${sessionId}/share`)
    return res.data
  },

  async revokeShare(sessionId: string): Promise<void> {
    await apiClient.delete(`/studio/sessions/${sessionId}/share`)
  },

  async exportPrintableReport(sessionId: string): Promise<string> {
    const res = await apiClient.get<string>(`/studio/sessions/${sessionId}/export`, {
      responseType: 'text',
    })
    return res.data
  },

  async getSharedSession(token: string): Promise<SharedStudioSession> {
    const res = await apiClient.get<SharedStudioSession>(`/studio/shared/${token}`)
    return res.data
  },

  sharedArtifactUrl(token: string, path: string): string {
    return `/api/v1/studio/shared/${encodeURIComponent(token)}/artifacts?path=${encodeURIComponent(path)}`
  },

  async extractSkill(
    sessionId: string,
    payload: ExtractStudioSkillPayload,
  ): Promise<ExtractedSkill> {
    const res = await apiClient.post<ExtractedSkill>(
      `/studio/sessions/${sessionId}/skills/extract`,
      payload,
    )
    return res.data
  },

  /** 完整保存工作区文本文件（编辑器 Ctrl/Cmd+S） */
  async writeFile(
    sessionId: string,
    payload: { path: string; content: string },
  ): Promise<{ path: string; size: number }> {
    const res = await apiClient.put<{ path: string; size: number }>(
      `/studio/sessions/${sessionId}/files/write`,
      payload,
    )
    return res.data
  },

  async makeDirectory(sessionId: string, path: string): Promise<void> {
    await apiClient.post(`/studio/sessions/${sessionId}/files/mkdir`, { path })
  },
  async renameFile(sessionId: string, path: string, newPath: string): Promise<void> {
    await apiClient.post(`/studio/sessions/${sessionId}/files/rename`, { path, new_path: newPath })
  },
  async deleteFile(sessionId: string, path: string): Promise<void> {
    await apiClient.delete(`/studio/sessions/${sessionId}/files/delete`, { params: { path } })
  },

  /** 精确编辑工作区文件（拒绝 workspace_edit 时交换 old/new 回滚） */
  async editFile(
    sessionId: string,
    payload: { path: string; old_string: string; new_string: string },
  ): Promise<StudioFileEditResult> {
    const res = await apiClient.post<StudioFileEditResult>(
      `/studio/sessions/${sessionId}/files/edit`,
      payload,
    )
    return res.data
  },

  /** 工作区目录列表（path 相对 /workspace） */
  async listFiles(
    sessionId: string,
    path = '',
  ): Promise<{ path: string; entries: WorkspaceEntry[] }> {
    const res = await apiClient.get<{ path: string; entries: WorkspaceEntry[] }>(
      `/studio/sessions/${sessionId}/files`,
      { params: { path } },
    )
    return res.data
  },

  /** 以 blob 方式读取任意受路径守卫保护的工作区文件（图片/二进制预览）。 */
  async fetchWorkspaceBlob(sessionId: string, path: string): Promise<string> {
    const res = await apiClient.get(`/studio/sessions/${sessionId}/files/download`, {
      params: { path },
      responseType: 'blob',
    })
    return URL.createObjectURL(res.data as Blob)
  },

  async downloadWorkspaceFile(sessionId: string, path: string): Promise<void> {
    const url = await studioApi.fetchWorkspaceBlob(sessionId, path)
    triggerBlobDownload(url, path.split('/').pop() || 'workspace-file')
  },

  /** 分页读取工作区文件 */
  async readFile(
    sessionId: string,
    path: string,
    offset = 0,
    limit = 200,
  ): Promise<{ content: string; total_lines: number; truncated: boolean }> {
    const res = await apiClient.get<{ content: string; total_lines: number; truncated: boolean }>(
      `/studio/sessions/${sessionId}/files/read`,
      { params: { path, offset, limit } },
    )
    return res.data
  },

  /** 将数据管理文件以只读软链引入 /workspace/input */
  async importDataFile(
    sessionId: string,
    payload: { file_id: string; name?: string },
  ): Promise<StudioImportResult> {
    const res = await apiClient.post<StudioImportResult>(
      `/studio/sessions/${sessionId}/import`,
      payload,
    )
    return res.data
  },

  /** 读取产物文本预览，实际复用工作区分页读取守卫 */
  async readArtifactText(
    sessionId: string,
    path: string,
    limit = 500,
  ): Promise<{ content: string; total_lines: number; truncated: boolean }> {
    return studioApi.readFile(sessionId, path, 0, limit)
  },

  /** 产物清单（/workspace/output） */
  async listArtifacts(sessionId: string): Promise<StudioArtifact[]> {
    const res = await apiClient.get<{ artifacts: StudioArtifact[] }>(
      `/studio/sessions/${sessionId}/artifacts`,
    )
    return res.data?.artifacts || []
  },

  /** 将 output/ 产物登记到报告中心；来源报告会话自动挂接版本树 */
  async registerArtifact(
    sessionId: string,
    payload: { path: string; title: string; type?: string; description?: string },
  ): Promise<StudioArtifactRegistration> {
    const res = await apiClient.post<StudioArtifactRegistration>(
      `/studio/sessions/${sessionId}/artifacts/register`,
      payload,
    )
    return res.data
  },

  /** 以 blob 方式拉取产物（携带 JWT），返回 Object URL，调用方负责 revoke */
  async fetchArtifactBlob(sessionId: string, path: string): Promise<string> {
    const res = await apiClient.get(`/studio/sessions/${sessionId}/artifacts/download`, {
      params: { path },
      responseType: 'blob',
    })
    return URL.createObjectURL(res.data as Blob)
  },

  /** 触发浏览器下载产物 */
  async downloadArtifact(sessionId: string, path: string): Promise<void> {
    const url = await studioApi.fetchArtifactBlob(sessionId, path)
    triggerBlobDownload(url, path.split('/').pop() || 'artifact')
  },
}

File: src/omichub/application/schemas/studio.py
"""OmicStudio AI 分析工作台 Pydantic DTO"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field

from omichub.application.schemas.chat import ChatSessionDTO


class CreateStudioSessionRequest(BaseModel):
    """创建 Studio 会话请求"""

    agent_id: str = Field(..., description="绑定的 Agent ID")
    title: str | None = Field(None, description="会话标题，缺省为「与 <Agent名> 的工作台」")
    project_id: str | None = Field(None, max_length=64, description="项目边界标识")
    model_id: UUID | None = Field(None, description="模型配置 ID，缺省用 Agent 绑定模型")


class PromoteStudioSessionRequest(BaseModel):
    """普通会话升级为 Studio 工作台请求（保留历史消息）"""

    agent_id: str | None = Field(
        None, description="可选：将会话重绑定到该 Agent（路由场景缺省沿用会话绑定 Agent）"
    )


class CreateStudioSessionFromReportRequest(BaseModel):
    """从报告创建 Studio 会话请求（结果报告中心「在 AI 工作台中优化」，§7.2）"""

    report_id: UUID = Field(..., description="来源报告 ID")
    agent_id: str = Field(..., description="绑定的 Agent ID")


class RegisterStudioArtifactRequest(BaseModel):
    """产物登记请求（右栏「保存为新版本」，与 artifact_register 工具同服务路径）"""

    path: str = Field(..., description="产物路径（相对 /workspace，必须在 output/ 下）")
    title: str = Field(..., description="报告标题")
    type: str | None = Field(None, description="产物类型（html/pdf/png/csv 等），缺省按扩展名推断")
    description: str | None = Field(None, description="报告描述")


class RegisterStudioArtifactResponse(BaseModel):
    """产物登记响应"""

    report_id: UUID
    title: str
    version: int = Field(..., description="版本号（挂接版本树时为父版本+1）")
    parent_id: UUID | None = Field(None, description="父报告 ID（版本树）")
    file: dict[str, Any] = Field(default_factory=dict, description="登记的文件信息")


class ImportStudioDataFileRequest(BaseModel):
    """从数据管理或聊天上传记录引入文件请求。"""

    file_id: str = Field(..., description="file://{uuid} 数据管理文件或 upload://{hex} 聊天上传文件")
    name: str | None = Field(None, description="工作区内显示名，缺省用原文件名")


class StudioPathRequest(BaseModel):
    path: str = Field(..., min_length=1, description="相对 /workspace 的路径")


class RenameStudioFileRequest(BaseModel):
    path: str = Field(..., min_length=1)
    new_path: str = Field(..., min_length=1)


class SaveStudioFileRequest(BaseModel):
    """从工作区编辑器完整保存 UTF-8 文本文件。"""

    path: str = Field(..., min_length=1, description="相对 /workspace 的文件路径")
    content: str = Field(..., max_length=1_048_576, description="完整文件内容，最大 1 MiB")


class SaveStudioFileResponse(BaseModel):
    """工作区文件保存结果。"""

    path: str
    size: int = 0


class EditStudioFileRequest(BaseModel):
    """Studio 文件编辑请求（workspace_edit 拒绝时用 old/new 交换回滚）。"""

    path: str = Field(..., description="相对 /workspace 的文件路径")
    old_string: str = Field(..., description="必须唯一出现的原文本")
    new_string: str = Field(..., description="替换文本")


class EditStudioFileResponse(BaseModel):
    """文件编辑结果，diff 为统一 diff 文本。"""

    path: str
    diff: str = ""
    size: int = 0


class ImportStudioDataFileResponse(BaseModel):
    """引入结果：沙盒内路径 + 当前 input/ 清单（与 datahub_import 工具 payload 同构）"""

    sandbox_path: str = Field(..., description="沙盒内可读路径，如 /workspace/input/a.csv")
    name: str = Field(..., description="工作区内链接名（重名自动加序号）")
    size: int = Field(..., description="文件大小（字节）")
    file_type: str = Field(..., description="数据管理或聊天上传文件类型")
    input_files: list[str] = Field(
        default_factory=list, description="当前 /workspace/input/ 下全部已引入路径"
    )


class UpdateStudioUiRequest(BaseModel):
    """会话级代码工作室视图状态。"""

    view_mode: Literal["chat", "split", "code"] | None = None
    split_ratio: float | None = Field(None, ge=25, le=70)
    follow_ai: bool | None = None
    terminal_collapsed: bool | None = None


class UpdateStudioPermissionsRequest(BaseModel):
    """会话级权限模式（监督 / 计划 / 放权）。"""

    mode: Literal["supervised", "plan", "auto"] = Field(..., description="权限模式")


class StudioApprovalApproveRequest(BaseModel):
    """批准工具审批；modified_args 为用户编辑后的参数（编辑后批准）。"""

    modified_args: dict[str, Any] | None = Field(None, description="编辑后的工具参数")


class StudioApprovalRejectRequest(BaseModel):
    """退回工具审批；reason 回灌 LLM 让其修改后重试。"""

    reason: str | None = Field(None, description="退回理由")


class StudioShareStatusDTO(BaseModel):
    active: bool
    expires_at: datetime | None = None
    shared_at: datetime | None = None


class StudioSessionDetailDTO(BaseModel):
    """Studio 会话详情：会话本体 + 沙盒状态 + 工作区文件树 + 当前计划"""

    session: ChatSessionDTO
    sandbox_status: str = Field(..., description="running / stopped / absent / unavailable")
    workspace_id: str | None = None
    files: list[dict[str, Any]] = Field(default_factory=list, description="工作区根目录条目")
    plan: dict[str, Any] | None = Field(None, description="最新待办计划（sandbox_meta.plan）")
    capabilities: dict[str, Any] | None = Field(
        None, description="会话级 MCP/Skill 加载状态与最近审计记录（sandbox_meta.capabilities）"
    )
    ui: dict[str, Any] = Field(default_factory=dict, description="代码工作室视图与交互设置")
    permissions: dict[str, Any] = Field(
        default_factory=lambda: {"mode": "supervised"},
        description="会话级权限模式（sandbox_meta.permissions，默认 supervised）",
    )
    sandbox_metrics: dict[str, Any] = Field(default_factory=dict, description="容器 CPU/内存快照")
    share: StudioShareStatusDTO | None = Field(
        None, description="只读分享状态；原始令牌永不从详情接口返回"
    )


class CreateStudioShareRequest(BaseModel):
    """创建或轮换 Studio 会话只读分享链接。"""

    expires_hours: int = Field(
        168, ge=1, le=720, description="分享有效期小时数，默认 7 天，最长 30 天"
    )


class StudioShareResponse(BaseModel):
    token: str = Field(..., description="仅创建/轮换时返回一次的高熵分享令牌")
    share_path: str = Field(..., description="公开只读页面路径")
    expires_at: datetime


class SharedStudioMessageDTO(BaseModel):
    role: str
    content: str
    created_at: datetime


class SharedStudioSessionDTO(BaseModel):
    title: str
    agent_id: str | None = None
    created_at: datetime
    updated_at: datetime
    expires_at: datetime
    messages: list[SharedStudioMessageDTO] = Field(default_factory=list)
    artifacts: list[dict[str, Any]] = Field(default_factory=list)


class ExtractStudioSkillRequest(BaseModel):
    """管理员将工作区脚本提炼为可复用 Skill。"""

    path: str = Field(..., min_length=1, max_length=240, description="工作区内脚本路径")
    skill_id: str = Field(..., min_length=1, max_length=50, pattern=r"^[a-z0-9][a-z0-9_-]{0,49}$")
    name: str = Field(..., min_length=1, max_length=100)
    description: str = Field("", max_length=500)
    usage_instructions: str = Field("", max_length=4000)
    category: str = Field("studio", max_length=50)
    icon: str = Field("🧪", max_length=10)
    confirm_no_secrets: bool = Field(
        False, description="管理员已人工检查脚本，不包含 API Key、密码或用户隐私数据"
    )


class StudioRunRequest(BaseModel):
    """用户重跑代码卡片请求（对应 /studio/sessions/{id}/run）"""

    code: str = Field(..., description="要执行的完整代码")
    language: Literal["python", "r", "bash"] = "python"
    timeout: int | None = Field(None, description="超时秒数，缺省用沙盒默认（600s）")


class StudioInternalExecRequest(BaseModel):
    """Worker -> Web 的内部沙盒执行请求。"""

    user_id: str
    language: Literal["python", "r", "bash"]
    code: str
    timeout_sec: int = Field(ge=1, le=3600)
    image: str | None = None
